doc/decisions/auth.md

```diff
@@ -0,0 +1,82 @@
+The users need to authenticate themself (who is the user).
+Then they get the appropriate authorization (does the user have the right privileges) to perform the action.
+
+**OAuth2.0:** **User** can authorize **client** to fetch resources from **resource server**.
+
+**OpenID Connect:** Extension of OAuth2.0 for authentication
+
+## Constraints
+
+- single account to access PermaplanT and Nextcloud
+- we don't implement our own auth solution
+
+## Assumptions
+
+- Nextcloud and PermaplanT backend are the only services we provide
+- we don't need roles/scopes (everyone can access all parts of the API)
+- access control is quite simple, e.g., which users can read or write on which map
+
+## Solutions
+
+### Alternative A
+OpenID Connect authentication & authorization flow with:
+
+Authorization server: [Nextcloud OIDC App](https://github.com/H2CK/oidc)
+
+Resource server 1: [Nextcloud OAuth2](https://docs.nextcloud.com/server/latest/admin_manual/configuration_server/oauth2.html)
+
+Resource server 2: PermaplanT backend
+
+Client: PermaplanT frontend
+
+**Notes:**
+> Currently it is not yet possible to use an issued Access Token or ID Token to access resources at the Nextcloud instance it self.
+
+\- [H2CK/oidc](https://github.com/H2CK/oidc)
+
+This is a major problem with this approach.
+
+> Nextcloud OAuth2 implementation currently does not support scoped access.
+> This means that every token has
+> full access to the complete account including read and write permission to the stored files.
+> It is essential to store the OAuth2 tokens in a safe way!
+> Without scopes and restrictable access it is
+> not recommended to use a Nextcloud instance as a user authentication service.
+
+\- [Nextcloud documentation](https://docs.nextcloud.com/server/latest/admin_manual/configuration_server/oauth2.html)
+
+### Alternative B
+[App password login flow](https://docs.nextcloud.com/server/latest/developer_manual/client_apis/LoginFlow/index.html)
+
+This is the login flow described in the Nextcloud documentation.
+A client can retrieve an app password for a user.
+Opposed to token authentication described before the app password
+is permanent and should not be stored anywhere in the frontend.
+This approach requires PermaplanT to have its own authentication server
+and therefore seperate accounts!
+
+### Alternative C
+OpenID Connect authentication & authorization flow with:
+
+Authorization server: PermaplanT
+
+Resource server 1: Nextcloud
+
+Resource server 2: PermaplanT
+
+Client 1: Nextcloud WebClient [with oidc login](https://github.com/pulsejet/nextcloud-oidc-login) or [user\_oidc](https://github.com/nextcloud/user_oidc)
+
+Client 2: PermaplanT
+
+**Notes:**
+In this approach the accounts are managed by PermaplanT,
+therefore requires implementing an identity provider in PermaplanT.
+Alternatively to developing an identity provider a service like [keycloak](https://www.keycloak.org/) can be used.
```

Comment:
```suggestion

## Decision

We will use [Keycloak](https://www.keycloak.org/).

```

@kitzbergerg can you give some info about how we would use Keycloak? (Do we need adapters, how to configure it etc. pp.)

Reply:
For our usecases I don't think we need any custom code/logic. 
Keycloak itself would run as a separate service like Nextcloud. You create users there instead of in Nextcloud/PermaplanT and users can then log in via Keycloak. 

So in the PermaplanT frontend you would get a redirect to Keycloak, you login in Keycloak and then you get redirected back to PermaplanT. From then on all requests are authorized via JWT. How that happens in detail is specified by the OAuth2 spec and there are also multiple ways of doing that (I don't think we need to worry about that in detail just use the recommended variant which is [Authorization Code Flow with PKCE](https://auth0.com/docs/get-started/authentication-and-authorization-flow/authorization-code-flow-with-proof-key-for-code-exchange-pkce)).
The backend you then simply validate the tokens and extract roles/user information out of them. There are probably libraries for that (although doing that part by hand is not that complicated).

Keycloak itself can either run as an executable or as a docker container. I have used it as a docker container (which would also be my preferred solution as configuration is easier that way). Apart from that it only requires a database so we can use our existing PostgreSQL for that.

Reply:
Thx, good suggestions, @absurd-turtle can you add this to the decision. (Except of Docker, we will probably use Ansible for deployment but this is not really relevant for this decision here.)